File: src/Plotly.NET/Traces/ObjectAbstractions/Selection.fs
namespace Plotly.NET.TraceObjects

open Plotly.NET
open Plotly.NET.LayoutObjects
open DynamicObj
open System
open System.Runtime.InteropServices

type MarkerSelectionStyle() =
    inherit ImmutableDynamicObj ()

    static member init
        (    
            [<Optional;DefaultParameterValue(null)>] ?Opacity:   float,
            [<Optional;DefaultParameterValue(null)>] ?Color:     Color,
            [<Optional;DefaultParameterValue(null)>] ?Size:      float
        ) =    
            MarkerSelectionStyle()
            |> MarkerSelectionStyle.style
                (
                    ?Opacity = Opacity,
                    ?Color   = Color,
                    ?Size    = Size
                )

    static member style
        (    
            [<Optional;DefaultParameterValue(null)>] ?Opacity:   float,
            [<Optional;DefaultParameterValue(null)>] ?Color:     Color,
            [<Optional;DefaultParameterValue(null)>] ?Size:      float
        ) =
            (fun (markerSelectionStyle:MarkerSelectionStyle) -> 

                markerSelectionStyle

                ++? ("opacity", Opacity)
                ++? ("color", Color)
                ++? ("size", Size)
            )

type FontSelectionStyle() =
    inherit ImmutableDynamicObj ()

    /// Init Font()
    static member init
        (    
            [<Optional;DefaultParameterValue(null)>] ?Color:     Color
        ) =    
            FontSelectionStyle()
            |> FontSelectionStyle.style
                (
                    ?Color  = Color
                )


    // Applies the styles to Font()
    static member style
        (    
            [<Optional;DefaultParameterValue(null)>] ?Color:     Color
        ) =
            (fun (fontSelectionStyle:FontSelectionStyle) ->  Color
                ++? ("color",) 
            )

type Selection () =
    inherit ImmutableDynamicObj ()

    static member init
        (    
            [<Optional;DefaultParameterValue(null)>] ?MarkerSelectionStyle: MarkerSelectionStyle,
            [<Optional;DefaultParameterValue(null)>] ?FontSelectionStyle: FontSelectionStyle
        ) =    
            Selection()
            |> Selection.style
                (
                    ?MarkerSelectionStyle   = MarkerSelectionStyle,
                    ?FontSelectionStyle     = FontSelectionStyle
                )

    static member style
        (    
            [<Optional;DefaultParameterValue(null)>] ?MarkerSelectionStyle: MarkerSelectionStyle,
            [<Optional;DefaultParameterValue(null)>] ?FontSelectionStyle: FontSelectionStyle
        ) =
            (fun (selection:Selection) -> 
                selection
                ++? ("marker", MarkerSelectionStyle)
                ++? ("textfont", FontSelectionStyle)
            )

File: src/Plotly.NET.CSharp/ObjectAbstractions/TraceObjects/Selection.cs

using Microsoft.FSharp.Core;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace Plotly.NET.CSharp;

public static partial class TraceObjects
{
    public static Plotly.NET.TraceObjects.MarkerSelectionStyle InitMarkerSelectionStyle
        (
            Optional<double> Opacity = default,
            Optional<Color> Color = default,
            Optional<double> Size = default
        ) 
            =>
                Plotly.NET.TraceObjects.MarkerSelectionStyle.init(
                    Opacity: Opacity.ToOption(),
                    Color: Color.ToOption(),
                    Size: Size.ToOption()
                );
    public static Plotly.NET.TraceObjects.FontSelectionStyle InitFontSelectionStyle
        (
            Optional<Color> Color = default
        ) 
            =>
                Plotly.NET.TraceObjects.FontSelectionStyle.init(
                    Color: Color.ToOption()
                );
    public static Plotly.NET.TraceObjects.Selection InitSelection
        (
            Optional<Plotly.NET.TraceObjects.MarkerSelectionStyle> MarkerSelectionStyle = default,
            Optional<Plotly.NET.TraceObjects.FontSelectionStyle> FontSelectionStyle = default
        ) 
            =>
                Plotly.NET.TraceObjects.Selection.init(
                    MarkerSelectionStyle: MarkerSelectionStyle.ToOption(),
                    FontSelectionStyle: FontSelectionStyle.ToOption()
                );
}

